Update 0011-NP-CLIE-client-interaction-tests.md (#1542)
vegaprotocol/specs: non-protocol-specs/0011-NP-CLIE-client-interaction-tests.md

    - This is communicated to the wallet, and wallet uses the new parameter for each transaction tied to a block with a height higher than the one in which the change happened. This requires the wallet to be subjected to a difficulty increase due to too many messages (<a name="0011-NP-CLIE-005" href="#0011-NP-CLIE-005">0011-NP-CLIE-005</a>)
6. The parameter `spam.pow.increaseDifficulty` is changed from `1` to `0`.  Verify that
    - This is communicated to the wallet, and wallet uses the new parameter for each transaction tied to a block with a height higher than the one in which the change happened. The wallet should submit no more than `spam.pow.numberofTxPerBlock` PoWs tied to a given block before using another. If this results in none being available the transaction should not be sent. (<a name="0011-NP-CLIE-006" href="#0011-NP-CLIE-006">0011-NP-CLIE-006</a>)

> 7. Link more than <SpamPoWNumberOfTxPerBlock>. (currently, 2) transactions to a single block for the PoW; assure that the client does not allow that and the account does not get blocked. (<a name="0011-NP-CLIE-007" href="#0011-NP-CLIE-007">0011-NP-CLIE-007</a>)

Reviewer: "Link more than <SpamPoWNumberOfTxPerBlock>" doesn't work from the client point of view. 

The whole PoW should be largely transparent to a client. 

A test can be more along the lines: set `spam.pow.numberOfPastBlocks = 100`. Take a client who sends 400 transactions as quickly as possible to the wallet. The wallet will correctly compute what PoW difficulty must be attached to which block historical block and submits these as quickly as possible without getting the party banned.

Also, < xxx > doesn't do any markdown formatting (since you wanted a brutal review).

> 7. Link more than <SpamPoWNumberOfTxPerBlock>. (currently, 2) transactions to a single block for the PoW; assure that the client does not allow that and the account does not get blocked. (<a name="0011-NP-CLIE-007" href="#0011-NP-CLIE-007">0011-NP-CLIE-007</a>)

Reviewer 2: ```suggestion
7. Link more than `SpamPoWNumberOfTxPerBlock`. (currently, 2) transactions to a single block for the PoW; assure that the client does not allow that and the account does not get blocked. (<a name="0011-NP-CLIE-007" href="#0011-NP-CLIE-007">0011-NP-CLIE-007</a>)
```

> 8. Submit significantly (i.e., twice) more than <SpamPoWNumberOfTxPerBlock> * <SpamPowNumberOfPastBlocks> (currently 200) transactions at once. Assure that the client delays or removes the transactions to not exceed the transactions-per-block limit and gets banned.(<a name="0011-NP-CLIE-008" href="#0011-NP-CLIE-008">0011-NP-CLIE-008</a>)

Reviewer 2: ```suggestion
8. Submit significantly (i.e., twice) more than `SpamPoWNumberOfTxPerBlock * SpamPowNumberOfPastBlocks` (currently 200) transactions at once. Assure that the client delays or removes the transactions to not exceed the transactions-per-block limit and gets banned.(<a name="0011-NP-CLIE-008" href="#0011-NP-CLIE-008">0011-NP-CLIE-008</a>)
```

> 9. Submit significantly (i.e., twice) more than <SpamPoWNumberOfTxPerBlock> * <SpamPowNumberOfPastBlocks> (currently 200) transactions at once, and keep submitting transactions at a rate of 20tx per second for at least a 10 minutes. Assure that the client delays or removes the transactions to not exceed the transactions-per-block limit and gets banned, and that the client does not crash due to a buffer overload.(<a name="0011-NP-CLIE-009" href="#0011-NP-CLIE-009">0011-NP-CLIE-009</a>)

Reviewer 2: ```suggestion
9. Submit significantly (i.e., twice) more than `SpamPoWNumberOfTxPerBlock * SpamPowNumberOfPastBlocks` (currently 200) transactions at once, and keep submitting transactions at a rate of 20 tx per second for at least a 10 minutes. Assure that the client delays or removes the transactions to not exceed the transactions-per-block limit and gets banned, and that the client does not crash due to a buffer overload.(<a name="0011-NP-CLIE-009" href="#0011-NP-CLIE-009">0011-NP-CLIE-009</a>)
```

> 10. Submit significantly (ten times) as many votes than allowed (<SpamProtectionMaxVotes>=3) for the same issue at the same time, and verify that the wallet does not allow more than (<SpamProtectionMaxVotes>=3) to be simultaneously, i.e., cause the account to get blocked.(<a name="0011-NP-CLIE-010" href="#0011-NP-CLIE-010">0011-NP-CLIE-010</a>)

Reviewer 2: ```suggestion
10. Submit significantly (ten times) as many votes than allowed (`SpamProtectionMaxVotes` = 3) for the same issue at the same time, and verify that the wallet does not allow more than (`SpamProtectionMaxVotes` = 3) to be simultaneously, i.e., cause the account to get blocked.(<a name="0011-NP-CLIE-010" href="#0011-NP-CLIE-010">0011-NP-CLIE-010</a>)
```

> 11. Submit significantly (ten times) as many proposals than allowed (<SpamProtectionMaxProposals>=3)at the same time (from an account that has more than <SpamProtectionMinTokensForProposal> tokens), and verify that the wallet does not allow more than (<SpamProtectionMaxProposals>=3) to be simultaneously, i.e., cause the account to get blocked.(<a name="0011-NP-CLIE-011" href="#0011-NP-CLIE-011">0011-NP-CLIE-011</a>)

Reviewer 2: ```suggestion
11. Submit significantly (ten times) as many proposals than allowed (`SpamProtectionMaxProposals` = 3)at the same time (from an account that has more than `SpamProtectionMinTokensForProposal` tokens), and verify that the wallet does not allow more than (`SpamProtectionMaxProposals` = 3) to be simultaneously, i.e., cause the account to get blocked.(<a name="0011-NP-CLIE-011" href="#0011-NP-CLIE-011">0011-NP-CLIE-011</a>)
```

> 12. Submit significantly (ten times) as many delegations than allowed (<SpamProtectionMaxDelegations>=390)at the same time (from an account that has more than <SpamProtectionMinTokensForDelegation> tokens), and verify that the wallet does not allow more than (<SpamProtectionMaxProposals>=390) to be simultaneously, i.e., cause the account to get blocked.(<a name="0011-NP-CLIE-012" href="#0011-NP-CLIE-012">0011-NP-CLIE-012</a>)

Reviewer 3: Minor: if I remember correctly, the spam rules for "delegations" include both delegating and undelegating so it might we both throwing at core half of one and half of the other.

Edit: You've used the phrase `delegation changes` below which is a nice way of putting it.

> 12. Submit significantly (ten times) as many delegations than allowed (<SpamProtectionMaxDelegations>=390)at the same time (from an account that has more than <SpamProtectionMinTokensForDelegation> tokens), and verify that the wallet does not allow more than (<SpamProtectionMaxProposals>=390) to be simultaneously, i.e., cause the account to get blocked.(<a name="0011-NP-CLIE-012" href="#0011-NP-CLIE-012">0011-NP-CLIE-012</a>)

Reviewer 2: ```suggestion
12. Submit significantly (ten times) as many delegation changes than allowed (`SpamProtectionMaxDelegations` = 390)at the same time (from an account that has more than `SpamProtectionMinTokensForDelegation` tokens), and verify that the wallet does not allow more than (`SpamProtectionMaxProposals` = 390) to be simultaneously, i.e., cause the account to get blocked.(<a name="0011-NP-CLIE-012" href="#0011-NP-CLIE-012">0011-NP-CLIE-012</a>)
```

> 13. Submit one less than (<SpamProtectionMaxVotes>=3) , i.e., 2 votes in the same epoch.  Once all votes  are processed, submit 2 new votes at once (still in the same epoch).  Verify that the wallet does not allow more than 1 to be submitted simultaneously, i.e., cause the account to get blocked. Submit 2 new simultaneous votes on that issue in the next epoch and verify that the wallet allows it (and the account does not get blocked)(<a name="0011-NP-CLIE-012" href="#0011-NP-CLIE-013">0011-NP-CLIE-013</a>)

Reviewer 2: ```suggestion
13. Submit one less than (`SpamProtectionMaxVotes `= 3) , i.e., 2 votes in the same epoch.  Once all votes  are processed, submit 2 new votes at once (still in the same epoch).  Verify that the wallet does not allow more than 1 to be submitted simultaneously, i.e., cause the account to get blocked. Submit 2 new simultaneous votes on that issue in the next epoch and verify that the wallet allows it (and the account does not get blocked)(<a name="0011-NP-CLIE-012" href="#0011-NP-CLIE-013">0011-NP-CLIE-013</a>)
```

> 14. Submit one less than (<SpamProtectionMaxProposals>=3) , i.e., 2 proposals in an epoch from an account that has more than <SpamProtectionMinTokensForProposal> tokens). Once all proposals are processed, submit 2 new proposals at once (still in the same epoch).  Verify that the wallet does not allow more than 1 to be submitted simultaneously, i.e., cause the account to get blocked. Submit 2 new proposals at the same time in the next epoch and verify that the wallet allows it (and the account does not get blocked)(<a name="0011-NP-CLIE-014" href="#0011-NP-CLIE-014">0011-NP-CLIE-014</a>)

Reviewer 2: ```suggestion
14. Submit one less than (`SpamProtectionMaxProposals` = 3) , i.e., 2 proposals in an epoch from an account that has more than `SpamProtectionMinTokensForProposal` tokens). Once all proposals are processed, submit 2 new proposals at once (still in the same epoch).  Verify that the wallet does not allow more than 1 to be submitted simultaneously, i.e., cause the account to get blocked. Submit 2 new proposals at the same time in the next epoch and verify that the wallet allows it (and the account does not get blocked)(<a name="0011-NP-CLIE-014" href="#0011-NP-CLIE-014">0011-NP-CLIE-014</a>)
```

> 15. Submit one less than (<SpamProtectionMaxDelegations>=390) , i.e., 389 delegations in the same epoch from an account that has more than <SpamProtectionMinTokensForDelegation> tokens). Once all delegations are processed, submit 100 new delegations at once (still in the same epoch).  Verify that the wallet does not allow more than 1 to be submitted simultaneously, i.e., cause the account to get blocked. Submit 100 new delegations at once in the next epoch and verify that the wallet allows it (and the account does not get blocked).(<a name="0011-NP-CLIE-015" href="#0011-NP-CLIE-015">0011-NP-CLIE-015</a>)

Reviewer 2: ```suggestion
15. Submit one less than (`SpamProtectionMaxDelegations` = 390) , i.e., 389 delegations in the same epoch from an account that has more than `SpamProtectionMinTokensForDelegation` tokens). Once all delegations are processed, submit 100 new delegations at once (still in the same epoch).  Verify that the wallet does not allow more than 1 to be submitted simultaneously, i.e., cause the account to get blocked. Submit 100 new delegations at once in the next epoch and verify that the wallet allows it (and the account does not get blocked).(<a name="0011-NP-CLIE-015" href="#0011-NP-CLIE-015">0011-NP-CLIE-015</a>)
```

> 16. Create a key with sufficient tokens to be allowed proposals and delegations (i.e., more than <SpamProtectionMinTokensForDelegation> and <SpamProtectionMinTokensForProposal>. Submit random (but valid) votes, delegation changes, transactions and proposals in sets of at least 13 each at the same time, at a rate of one set per second. Let that run for 4 epochs, with an epoch length of at least 100 blocks. Verify that the account never gets banned, and that the client never crashed. (<a name="0011-NP-CLIE-016" href="#0011-NP-CLIE-016">0011-NP-CLIE-016</a>)

Reviewer 2: ```suggestion
16. Create a key with sufficient tokens to be allowed proposals and delegations (i.e., more than `SpamProtectionMinTokensForDelegation` and `SpamProtectionMinTokensForProposal`. Submit random (but valid) votes, delegation changes, transactions and proposals in sets of at least 13 each at the same time, at a rate of one set per second. Let that run for 4 epochs, with an epoch length of at least 100 blocks. Verify that the account never gets banned, and that the client never crashed. (<a name="0011-NP-CLIE-016" href="#0011-NP-CLIE-016">0011-NP-CLIE-016</a>)
```

> 16. Create 20 keys with sufficient tokens to be allowed proposals and delegations (i.e., more than <SpamProtectionMinTokensForDelegation> and <SpamProtectionMinTokensForProposal>. For all keys (within their respective wallets), submit random (but valid) votes, delegation changes, transactions and proposals in sets of at least 13 each at the same time, at a rate of one set per second. Let that run for 4 epochs, with an epoch length of at least 100 blocks. Verify that the no account ever gets banned, and that the clients never crash. (<a name="0011-NP-CLIE-017" href="#0011-NP-CLIE-017">0011-NP-CLIE-017</a>)

Reviewer 2: ```suggestion
16. Create 20 keys with sufficient tokens to be allowed proposals and delegations (i.e., more than `SpamProtectionMinTokensForDelegation` and `SpamProtectionMinTokensForProposal`. For all keys (within their respective wallets), submit random (but valid) votes, delegation changes, transactions and proposals in sets of at least 13 each at the same time, at a rate of one set per second. Let that run for 4 epochs, with an epoch length of at least 100 blocks. Verify that the no account ever gets banned, and that the clients never crash. (<a name="0011-NP-CLIE-017" href="#0011-NP-CLIE-017">0011-NP-CLIE-017</a>)
```
